### PR TITLE
Dex based authentication

### PR DESCRIPTION
**Pull request by bonifaido**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| API breaks?     | no
| Deprecations?   | yes
| Related tickets | partially fixes https://github.com/banzaicloud/pipeline/issues/1476
| License         | Apache 2.0

### What's in this PR?
This PR is an initial implementation of the Dex based authentication mechanism, which is very minimal at first, and tries to keep good compatibility with the previous direct GitHub based authentication.

### Why?
With the help of Dex we can support as many auth providers as Dex already supports without substantional changes in the Pipeline auth code base.

### Checklist

- [x] Implementation tested (with at least one cloud provider)
- [ ] Error handling code meets the [guideline](https://github.com/banzaicloud/pipeline/blob/master/docs/error-handling-guide.md)
- [ ] Logging code meets the guideline (TODO)
- [x] OpenAPI and Postman files updated, client regenerated (`make generate-client`) (if needed)
- [x] User guide and development docs updated (if needed)
- [x] Related Helm chart(s) updated (if needed)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -85,6 +85,9 @@ vendor: bin/dep ## Install dependencies
 config/config.toml:
 	cp config/config.toml.example config/config.toml
 
+config/dex-config.yml:
```
Comment (sagikazarmark) [inline Makefile]: If the name `dex-config.yml` is not mandatory, I would just call it `dex.yml`

File: api/user.go [hunk 2]
```diff
@@ -246,3 +246,63 @@ func (a *UserAPI) RemoveUser(c *gin.Context) {
...
+		message := "failed to bind update user request"
+		c.AbortWithStatusJSON(http.StatusBadRequest, common.ErrorResponse{
+			Code:    http.StatusInternalServerError,
+			Message: message,
+			Error:   message,
+		})
+		return
+	}
+
+	db := config.DB()
```
Comment (sagikazarmark) [inline api/user.go]: db should be injected into the UserAPI struct

File: api/user.go [hunk 3]
```diff
@@ -246,3 +246,63 @@ func (a *UserAPI) RemoveUser(c *gin.Context) {
 
 	c.Status(http.StatusNoContent)
 }
+
+type updateUserRequest struct {
+	GitHubToken string `json:"gitHubToken,omitempty"`
+}
+
+// UpdateCurrentUser responds with the authenticated user
```
Comment (sagikazarmark) [inline api/user.go]: Is this description correct?

File: api/user.go [hunk 4]
```diff
@@ -246,3 +246,63 @@ func (a *UserAPI) RemoveUser(c *gin.Context) {
...
+		return
+	}
+
+	var updateUserRequest updateUserRequest
+	err := c.BindJSON(&updateUserRequest)
+
+	if err != nil {
+		message := "failed to bind update user request"
+		c.AbortWithStatusJSON(http.StatusBadRequest, common.ErrorResponse{
+			Code:    http.StatusInternalServerError,
```
Comment (sagikazarmark) [inline api/user.go]:
```suggestion
			Code:    http.StatusBadRequest,
```

File: api/user.go [hunk 5]
```diff
@@ -246,3 +246,63 @@ func (a *UserAPI) RemoveUser(c *gin.Context) {
...
+		return
+	}
+
+	db := config.DB()
+
+	err = db.Find(user).Error
+
+	if err != nil {
+		message := "failed to fetch user"
+		log.Info(message + ": " + err.Error())
```
Comment (sagikazarmark) [inline api/user.go]: logger should be injected into the UserAPI struct

Comment (sagikazarmark) [reply api/user.go]: also, this is hardly an INFO level log. It is an error, it should be handled by an errorHandler.

File: api/user.go [hunk 6]
```diff
@@ -246,3 +246,63 @@ func (a *UserAPI) RemoveUser(c *gin.Context) {
...
+			Error:   message,
+		})
+		return
+	}
+
+	if updateUserRequest.GitHubToken != "" {
+		err = auth.SaveUserGitHubToken(user, updateUserRequest.GitHubToken)
+		if err != nil {
+			message := "failed to update user's github token"
+			log.Info(message + ": " + err.Error())
```
Comment (sagikazarmark) [inline api/user.go]: Same as above

File: auth/user.go [hunk 7]
```diff
@@ -207,34 +210,125 @@ type BanzaiUserStorer struct {
...
 	}
 
-	// This assumes GitHub auth only right now
-	githubExtraInfo := schema.RawInfo.(*GithubExtraInfo)
-	currentUser.Login = githubExtraInfo.Login
-	err = bus.createUserInCICDDB(currentUser, githubExtraInfo.Token)
+	organizations, err := getOrganizationsFromDex(schema)
 	if err != nil {
-		log.Info(context.Request.RemoteAddr, err.Error())
+		log.Errorln("failed to parse groups/organizations:", err)
```
Comment (sagikazarmark) [inline auth/user.go]: do not log and return an error, rather annotate it if necessary

File: auth/user.go [hunk 8]
```diff
@@ -207,34 +210,125 @@ type BanzaiUserStorer struct {
...
 
-	synchronizeCICDRepos(currentUser.Login)
+	// Until https://github.com/dexidp/dex/issues/1076 gets resolved we need to use a manual
+	// GitHub API query to get the user login and image to retain compatibility for now
+	var githubUserMeta *githubUserMeta
+	if schema.Provider == "dex:github" {
+		githubUserMeta, err = getGithubUserMeta(schema)
+
+		if err != nil {
+			log.Errorln("failed to query github login name:", err)
```
Comment (sagikazarmark) [inline auth/user.go]: do not log and return an error, rather annotate it if necessary

File: auth/user.go [hunk 9]
```diff
@@ -207,34 +210,125 @@ type BanzaiUserStorer struct {
...
+		// Login will be the email for new users coming from an other provider than GitHub
+		currentUser.Login = schema.Email
+	}
+
+	db := authCtx.Auth.GetDB(authCtx.Request)
+
+	if viper.GetBool("auth.whitelistEnabled") {
+
+		if ok, err := checkWhiteList(db, currentUser, schema, organizations); err != nil {
+			log.Errorln("failed to check whitelist:", err)
```
Comment (sagikazarmark) [inline auth/user.go]: do not log and return an error

File: auth/user.go [hunk 10]
```diff
@@ -207,34 +210,125 @@ type BanzaiUserStorer struct {
...
+			return nil, "", err
+		} else if !ok {
+			return nil, "", errors.New("user is not enabled")
+		}
+	}
+
+	// TODO we should call the Drone API instead and insert the token later on manually by the user
+	err = bus.createUserInCICDDB(currentUser)
+	if err != nil {
+		log.Info(authCtx.Request.RemoteAddr, err.Error())
```
Comment (sagikazarmark) [inline auth/user.go]: wrong level, do not log and return an error, rather annotate it if necessary

File: auth/user.go [hunk 11]
```diff
@@ -246,56 +340,54 @@ func (bus BanzaiUserStorer) Save(schema *auth.Schema, context *auth.Context) (us
...
-func (bus BanzaiUserStorer) Update(schema *auth.Schema, context *auth.Context) error {
-
-	// This assumes GitHub auth only right now
-	currentUser := &User{}
-	githubExtraInfo := schema.RawInfo.(*GithubExtraInfo)
-	currentUser.Login = githubExtraInfo.Login
-
-	// Revoke the old Github token from Vault
-	err := TokenStore.Revoke(context.Claims.UserID, GithubTokenID)
+// func (bus BanzaiUserStorer) Update(schema *auth.Schema, context *auth.Context) error {
```
Comment (sagikazarmark) [inline auth/user.go]:
```suggestion
```

File: auth/user.go [hunk 12]
```diff
@@ -246,56 +340,54 @@ func (bus BanzaiUserStorer) Save(schema *auth.Schema, context *auth.Context) (us
...
+// // This will run on every login
```
Comment (sagikazarmark) [inline auth/user.go]:
```suggestion
```

File: auth/user.go [hunk 13]
```diff
@@ -246,56 +340,54 @@ func (bus BanzaiUserStorer) Save(schema *auth.Schema, context *auth.Context) (us
...
+// }
```
Comment (sagikazarmark) [inline auth/user.go]:
```suggestion
```

File: auth/user.go [hunk 14]
```diff
@@ -348,8 +440,28 @@ func NewGithubImporter(
 	}
 }
 
-func (i *GithubImporter) ImportOrganizations(currentUser *User, githubToken string) error {
-	githubOrgIDs, err := importGithubOrganizations(i.db, currentUser, githubToken)
+func (i *GithubImporter) ImportOrganizationsFromGithub(currentUser *User, githubToken string) error {
+
```
Comment (sagikazarmark) [inline auth/user.go]:
```suggestion
```

File: auth/user.go [hunk 15]
```diff
@@ -348,8 +440,28 @@ func NewGithubImporter(
...
+	orgs, err := getGithubOrganizations(githubToken)
+	if err != nil {
+		return emperror.With(err, "failed to get organizations")
```
Comment (sagikazarmark) [inline auth/user.go]:
```suggestion
		return emperror.Wrap(err, "failed to get organizations")
```

File: docker-compose.yml [hunk 16]
```diff
@@ -106,3 +106,9 @@ services:
         environment:
             CLOUDINFO_URL: https://beta.banzaicloud.io/cloudinfo
             RECOMMENDER_URL: https://beta.banzaicloud.io/recommender
+
+    dex:
+        image: banzaicloud/dex-shim:master
+        command: serve /dex-config.yml
```
Comment (sagikazarmark) [inline docker-compose.yml]: How about just `dex.yml`?

File: Makefile (hunk 1 above)
Comment (bonifaido) [reply Makefile]: renamed, thanks

File: api/user.go (hunk 2 above)
Comment (bonifaido) [reply api/user.go]: fixed, thanks!

File: api/organization.go [hunk 17]
```diff
@@ -160,7 +160,16 @@ func (a *OrganizationAPI) SyncOrganizations(c *gin.Context) {
 		return
 	}
 
-	err = a.githubImporter.ImportOrganizations(user, token)
+	if token == "" {
+		c.JSON(http.StatusInternalServerError, common.ErrorResponse{
```
Comment (sagikazarmark) [inline api/organization.go]:
```suggestion
		c.JSON(http.StatusBadRequest, common.ErrorResponse{
```

File: auth/user.go [hunk 18]
```diff
@@ -246,56 +337,48 @@ func (bus BanzaiUserStorer) Save(schema *auth.Schema, context *auth.Context) (us
...
 	}
 
-	// Save the new Github token to Vault
 	token := bauth.NewToken(GithubTokenID, "Github access token")
-	token.Value = githubExtraInfo.Token
-	err = TokenStore.Store(context.Claims.UserID, token)
+	token.Value = githubToken
+	err = TokenStore.Store(user.IDString(), token)
+	if err != nil {
+		return errors.Wrapf(err, "failed to store Github access token for user: %s", user.Login)
```
Comment (sagikazarmark) [inline auth/user.go]:
```suggestion
		return errors.WrapWith(err, "failed to store Github access token for user", "user", user.Login)
```

File: auth/user.go [hunk 19]
```diff
@@ -246,56 +337,48 @@ func (bus BanzaiUserStorer) Save(schema *auth.Schema, context *auth.Context) (us
...
+	}
+
+	// TODO CICD should use Vault as well, and this should be removed by then
+	err = updateUserInCICDDB(user, githubToken)
 	if err != nil {
-		return errors.Wrap(err, "failed to save Github access token")
+		return errors.Wrapf(err, "failed to update Github access token for user in CICD: %s", user.Login)
```
Comment (sagikazarmark) [inline auth/user.go]:
```suggestion
		return errors.WrapWith(err, "failed to update Github access token for user in CICD", "user", user.Login)
```

File: auth/user.go (hunk 18 above)
Comment (sagikazarmark) [reply auth/user.go]: Fields will end up in the log if the error is handled by an error handler.

File: auth/user.go (hunk 19 above)
Comment (sagikazarmark) [reply auth/user.go]: Fields will end up in the log if the error is handled by an error handler.

**Pull request merged**